### ISSUE TITLE
Looking for guidance on defining CSP for <portal>

### ISSUE THREAD

**Issue by domenic** [w3c/webappsec-csp#437]
Hi CSP spec community,

Over in https://github.com/WICG/portals we're looking to define a new HTML element, `<portal>`, which is somewhere intermediate between an `<iframe>` and a `<link rel="prerender">`. An especially relevant section to read for this question might be [Summary of differences between portals and iframes](https://github.com/WICG/portals#summary-of-differences-between-portals-and-iframes).

I've created a draft plan for CSP integration at https://github.com/WICG/portals/pull/235, and would love review and comments from this group.

A particular question we're uncertain on is how portals should situate themselves with respect to `child-src`, `frame-src`, and `prefetch-src`. My draft introduces a new `portal-src` directive, and makes that part of `child-src`, but not `frame-src` or `prefetch-src`.

This seems like a reasonable design (at least from my non-CSP-expert vantage) if we were designing the platform from scratch. But we're wondering about the impact on incremental deployment. For example, would pages be vulnerable in this design, e.g. if XSS vulnerability allowed introducing `<portal>`s, and they slipped past the page's CSP since `frame-src` does not block it? Or, is the idea that people generally set `default-src` tightly, and use `frame-src` to loosen things up, in which case bundling `<portal>` into `frame-src` would actually be more likely to cause them to slip by?

Does the fact that `object-src` is separate from `frame-src` mean anything here? And how should we think about this with relation to `prefetch-src`? I notice that the latter isn't even captured by `child-src`...

Do the various restrictions on portals [to prevent cross-site tracking](https://github.com/WICG/portals#privacy-threat-model-and-restrictions) or [to prevent them from affecting content outside their rendering region](https://github.com/WICG/portals#other-restrictions-while-portaled) change the calculus at all? What is the use case and threat model for `prefetch-src` and `frame-src` anyway?

**Comment by devd** [comment]
(I wonder if the "community" discussion can happen better on the mailing list than github issue?)

I am not sure .. isn't portal closer to the proposed "navigate-to" directive than frame-src / child-src?

**Comment by domenic** [comment]
Indeed, navigate-to also applies to portals, for activation. The question is what should apply during their pre-activation state.

**Comment by devd** [comment]
whats an example where we will want portals to load but not allow activation?  I see some mention of ads in the spec but I couldn't quite understand the use case in the spec.

**Comment by domenic** [comment]
> whats an example where we will want portals to load but not allow activation?

Allowing a portal to load, but disallowing activation, is similar to allowing a `<link rel="prefetch">` to load (i.e., not prohibiting it with `prefetch-src`), but disallowing navigation to that link (i.e., using `navigate-to`).

I'm not really sure doing so is very sensible, but currently CSP has `prefetch-src` and `navigate-to` as orthogonal primitives, so probably it makes sense to preserve that for portals?

>  I see some mention of ads in the spec but I couldn't quite understand the use case in the spec.

I'm not sure that use case makes as much sense anymore, given that we're mostly focusing on prerendering.  But, the basic idea is/was that if you wanted ads with [no cross-site tracking capability](https://github.com/WICG/portals#privacy-threat-model-and-restrictions) and very tight [sandboxing](https://github.com/WICG/portals#other-restrictions-while-portaled), you could use portals instead of iframes for those. We should reevaluate whether anyone wants that; it certainly wasn't partner-driven.

**Comment by devd** [comment]
hmm .. prefetch is slightly different in that it could be any resource while portals are explicitly navigation only. With prefetch, doesn't know which one of {img, object, style, script}-src to apply. In this case, I think its clear to the browser that it should apply navigation-to. So, e.g., I feel like the comparison could be having prefetch-img-src and img-src being two different directives. 

(and ack thanks for the explanation of the ads/portals situation)
